enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -351,79 +351,67 @@ impl<'a> Handler<'a> {
-        // array, measured in bytes.
-        let mut offset = 0;
-        for (t, mut u) in trusted.iter_mut().zip(untrusted.iter_mut()) {
-            u.iov_base = ubuffer[offset..].as_mut_ptr() as *mut _;
+        // Update pointers from iovecs to buffers in untrusted memory.
+        let mut size = 0;
+        for (t, u) in tiovec.iter().zip(uiovec.iter_mut()) {
+            size += t.iov_len;
             u.iov_len = t.iov_len;
-            offset += t.iov_len;
+            u.iov_base = match unsafe { cursor.alloc(t.iov_len) } {
+                Ok(base) => base.as_mut_ptr(),
+                Err(_) => return -libc::EMSGSIZE as u64,
+            };
         }
 
         // Do the syscall; replace encrypted memory with unencrypted memory.
-        let ret = unsafe {
-            self.syscall(
-                libc::SYS_readv as u64,
-                fd,
-                untrusted.as_ptr() as _,
-                untrusted.len() as u64,
-                0,
-                0,
-                0,
-            )
-        };
+        let req = request!(libc::SYS_readv => self.aex.gpr.rdi, uiovec, uiovec.len());
+        let res = unsafe { self.proxy(req) };
 
-        // Copy the unencrypted input into encrypted memory.
-        if ret <= ERRNO_BASE {
-            if ret > size as u64 {
-                self.attacked();
-            }
+        match res {
+            Ok(res) => {
+                if size < res[0].into() {
+                    self.attacked();
+                }
 
-            let mut offset = 0;
-            for (t, u) in trusted.iter_mut().zip(untrusted.iter_mut()) {
-                let us = &mut ubuffer[offset..][..t.iov_len];
-                offset += t.iov_len;
+                // Allocate the correct number of iovecs in untrusted memory.
+                // We reallocate again here to keep the Rust compiler happy.
+                let cursor = self.block.cursor();
+                let uiovec: &mut [libc::iovec] = match unsafe { cursor.alloc(tiovec.len()) } {
+                    Ok(slice) => slice,
+                    Err(_) => return -libc::EMSGSIZE as u64,
+                };
+
+                let mut bytes_copied = 0;
+                for (t, u) in tiovec.iter_mut().zip(uiovec.iter_mut()) {
+                    u.iov_len = t.iov_len;
+                    u.iov_base = match unsafe { cursor.alloc(t.iov_len) } {
+                        Ok(base) => base.as_mut_ptr(),
+                        Err(_) => return -libc::EMSGSIZE as u64,
+                    };
+
+                    // Grab data from untrusted buffers and copy to trusted buffers.
+                    let ubuf = unsafe { from_raw_parts(u.iov_base as *mut u8, u.iov_len) };
+                    let tbuf = unsafe { from_raw_parts_mut(t.iov_base as *mut u8, t.iov_len) };
+                    if bytes_copied < res[0].into() {
+                        tbuf.copy_from_slice(ubuf);
+                    }
```

Comment:
I'd invite you to compare this implementation with my version.

1. Rather than starting at zero and counting up I start at the bytes read and count down. This allows me to efficiently use `core::cmp::min()` to get the bytes to read, which reduces to zero once we've copied all the read bytes.

2. I never use the value of the untrusted iovecs. Because we handed this memory to the host, we'd have to validate it all for malicious modifications. Instead, I simply reproduce the slices as we did before the syscall. Because this is deterministic, we get the same results but don't have to validate the values.

Reply:
I like this approach.